storage/redis/peer_store.go

```diff
@@ -0,0 +1,803 @@
+
+	// creates Backend instance
+	h, p, db, err := ParseRedisURL(cfg.RedisBroker)
+	if err != nil {
+		return nil, err
+	}
+
+	ps := &peerStore{
+		cfg:    cfg,
+		rb:     NewBackend(h, p, "", db),
+		closed: make(chan struct{}),
+	}
+
+	// Start a goroutine for garbage collection.
+	ps.wg.Add(1)
+	go func() {
+		defer ps.wg.Done()
+		for {
+			select {
+			case <-ps.closed:
+				return
+			case <-time.After(cfg.GarbageCollectionInterval):
+				before := time.Now().Add(-cfg.PeerLifetime)
+				log.Debug("storage: purging peers with no announces since", log.Fields{"before": before})
+				ps.collectGarbage(before)
+			}
+		}
+	}()
+
+	// Start a goroutine for reporting statistics to Prometheus.
+	ps.wg.Add(1)
+	go func() {
+		defer ps.wg.Done()
+		t := time.NewTicker(cfg.PrometheusReportingInterval)
+		for {
+			select {
+			case <-ps.closed:
+				t.Stop()
+				return
+			case <-t.C:
+				before := time.Now()
+				ps.populateProm()
+				log.Debug("storage: populateProm() finished", log.Fields{"timeTaken": time.Since(before)})
+			}
+		}
+	}()
+
+	return ps, nil
+}
+
+type serializedPeer string
+
+func newPeerKey(p bittorrent.Peer) serializedPeer {
+	b := make([]byte, 20+2+len(p.IP.IP))
+	copy(b[:20], p.ID[:])
+	binary.BigEndian.PutUint16(b[20:22], p.Port)
+	copy(b[22:], p.IP.IP)
+
+	return serializedPeer(b)
+}
+
+func decodePeerKey(pk serializedPeer) bittorrent.Peer {
+	peer := bittorrent.Peer{
+		ID:   bittorrent.PeerIDFromString(string(pk[:20])),
+		Port: binary.BigEndian.Uint16([]byte(pk[20:22])),
+		IP:   bittorrent.IP{IP: net.IP(pk[22:])}}
+
+	if ip := peer.IP.To4(); ip != nil {
+		peer.IP.IP = ip
+		peer.IP.AddressFamily = bittorrent.IPv4
+	} else if len(peer.IP.IP) == net.IPv6len { // implies toReturn.IP.To4() == nil
+		peer.IP.AddressFamily = bittorrent.IPv6
+	} else {
+		panic("IP is neither v4 nor v6")
+	}
+
+	return peer
+}
+
+// Backend represents a Memcache result backend
+type Backend struct {
+	host     string
+	password string
+	db       int
+	pool     *redis.Pool
+	// If set, path to a socket file overrides hostname
+	socketPath string
+	redsync    *redsync.Redsync
+	common.RedisConnector
+}
+
+type peerStore struct {
+	cfg Config
+	rb  *Backend
+
+	closed chan struct{}
+	wg     sync.WaitGroup
+}
+
+// populateProm aggregates metrics over all shards and then posts them to
+// prometheus.
+func (ps *peerStore) populateProm() {
+	var numInfohashes, numSeeders, numLeechers uint64
+
+	shards := [2]string{bittorrent.IPv4.String(), bittorrent.IPv6.String()}
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	for _, shard := range shards {
+		ihList, err := conn.Do("HKEYS", shard) // key
+		if err != nil {
+			return
+		}
+		infohashes := ihList.([]interface{})
+
+		InfohashLPrefix := shard + "_L_"
+		InfohashSPrefix := shard + "_S_"
+		InfohashPrefixLen := len(InfohashLPrefix)
+		InfohashesMap := make(map[string]bool)
+
+		for _, ihKey := range infohashes {
+			ihStr := string(ihKey.([]byte))
+			ih := ihStr[InfohashPrefixLen:]
```

Comment:
This could panic if `len(ihStr) < InfohashPrefixLen`. Additionally, it is questionable to me why you take the first `InfohashPrefixLen` bytes of this string, to then check whether those have the `InfohashSPrefix` or `InfohashLPrefix` as prefix, instead of just comparing them OR doing the prefix check in the first place, without trimming.